Recoverable Transaction History Payment ID Memos (#54)
mobilecoinfoundation/mcips: text/0054-rth-payment-id-memos.md

- Feature Name: `Recoverable Transaction History Payment ID Memos`
- Start Date: (2022-10-18)
- MCIP PR: [mobilecoinfoundation/mcips#0054](https://github.com/mobilecoinfoundation/mcips/pull/54)

# Summary
[summary]: #summary

[#4 Recoverable Transaction History (RTH)](https://github.com/mobilecoinfoundation/mcips/pull/4) introduced three different memo types for [encrypted memos](https://github.com/mobilecoinfoundation/mcips/pull/3). This proposal specifies three more memo types:

- Destination With Payment Request Id Memo

> - Authenticated Sender With Payment Id Memo

Reviewer: Can we add the word `Intent` to these two memo names - Authenticated Sender With Payment Intent Id Memo, Destination With Payment Intent Id Memo - it will help disambiguate between a "request" and an "intent". Left simply as "Payment Id Memo" someone could conceivably mismatch with "Payment Request ID".

Author: Done

- Destination With Payment Id Memo

This feature will accomplish two main things:

1. Allow clients utilizing payment requests to recover complete transaction history for both users
2. Implement recoverable transaction history for payment intent flows

# Motivation
[motivation]: #motivation

The Authenticated Sender With Payment Request Id Memo was introduced to support clients implementing a payment request flow. A user can request a payment with a specific ID. Upon receiving a payment and sender memo with the same payment request ID, it can be reasonably assumed that the newly received payment is associated with the original request.

The user fulfilling a payment request this way, however, has no way of identifying any outgoing payment as belonging to a specific request ID after sending it. They only attach a regular destination memo to the change TxOut. Only the user which made the payment request is capable of reconstructing this aspect of the transaction history. This should be resolved so that recoverable transaction history between two parties is the same for both. This can be accomplished by creating a new destination memo type which contains the payment request ID.

> Additionally, some clients may wish to implement the opposite flow. A sending user could use an ID to enable the recipient to associate that payment with some previously shared and/or agreed upon payment intent. It is proposed that this be implemented with two new memo types: the Authanticated Sender With Payment Id Memo and the Destination with Payment Id Memo.

Reviewer: ```suggestion
Additionally, some clients may wish to implement the opposite flow. A sending user could use an ID to enable the recipient to associate that payment with some previously shared and/or agreed upon payment intent. It is proposed that this be implemented with two new memo types: the Authenticated Sender With Payment Id Memo and the Destination with Payment Id Memo.
```

# Guide-level explanation
[guide-level-explanation]: #guide-level-explanation

This proposal is an addition to [#4 Recoverable Transaction History](https://github.com/mobilecoinfoundation/mcips/pull/4). As such, the information included on new memo types here builds on the background recoverable transaction history information contained in that proposal.

## Destination With Payment Request Id Memo

> This memo will be paired with recoverable transaction history's Authenticated Sender With Paymend Request Id Memo. This memo will have type bytes `0x0203`. The Authenticated Sender With Paymend Request Id Memo that RTH defines is sent to a user to let them know that the incoming payment is related to a specific payment request. However, attached to the change TxOut of that transaction is just a normal Destination Memo. This prevents the user sending the transaction from having a record of fulfilling a payment request. The party receiving the transaction therefore has more information about the transaction history than the sender.

Reviewer: ```suggestion
This memo will be paired with recoverable transaction history's Authenticated Sender With Payment Request Id Memo. This memo will have type bytes `0x0203`. The Authenticated Sender With Payment Request Id Memo that RTH defines is sent to a user to let them know that the incoming payment is related to a specific payment request. However, attached to the change TxOut of that transaction is just a normal Destination Memo. This prevents the user sending the transaction from having a record of fulfilling a payment request. The party receiving the transaction therefore has more information about the transaction history than the sender.
```

# Reference-level explanation
[reference-level-explanation]: #reference-level-explanation

## 0x0203 Destination With Payment Request Id Memo

The 64-byte memo data is laid out as follows:

| Byte range | Item |
| ---------- | ---- |
| 0 - 16     | Recipient's Address Hash |
| 16 - 17    | The number of recipients, as an unsigned 8-bit number |
| 17 - 24    | Big-endian bytes of fee amount, as an unsigned 56-bit number |
| 24 - 32    | Big-endian bytes of the total outlay amount, as an unsigned 64-bit number |
| 32 - 40    | Big-endian bytes of 8-byte payment request id number |
| 40 - 64    | Unused bytes                  |

> This is exactly identical to RTH's destination memo except it contains an 8-byte payment request ID number. A user fulfilling a payment request will use this memo and populate that field with the same ID they put into an authenticated sender with payment request ID memo. This will allow the fulfilling user to keep a record that the payment request was fulfilled.

Reviewer 2: ```suggestion
This is RTH's 0x0200 Destination Memo plus an 8-byte payment request ID. A user fulfilling a payment request will use this memo and populate that field with the same ID they put into an Authenticated Sender With Payment Request ID Memo. This will allow the fulfilling user to keep a record that the payment request was fulfilled.
```

> ## 0x0102 Authenticated Sender With Payment Id Memo

Reviewer 2: Please move 0x0102 to the top

Author: Done

The 64-byte memo data is laid out as follows:

| Byte range | Item |
| ---------- | ---- |
| 0 - 16     | Sender's Address Hash |
| 16 - 24    | Big-endian bytes of 8-byte payment intent id number |
| 24 - 48    | Unused bytes          |
| 48 - 64    | HMAC                  |

This is exactly identical to RTH's authenticated sender with payment request ID memo. The only difference is the memo type. Utilizing the 2-byte memo type, client applications can differentiate payment requests from payment intents.

## 0x0204 Destination With Payment Id Memo

The 64-byte memo data is laid out as follows:

| Byte range | Item |
| ---------- | ---- |
| 0 - 16     | Recipient's Address Hash |
| 16 - 17    | The number of recipients, as an unsigned 8-bit number |
| 17 - 24    | Big-endian bytes of fee amount, as an unsigned 56-bit number |
| 24 - 32    | Big-endian bytes of the total outlay amount, as an unsigned 64-bit number |
| 32 - 40    | Big-endian bytes of 8-byte payment intent id number |
| 40 - 64    | Unused bytes                  |

> This is exactly identical to the herein proposed destination with payment request ID memo. The only difference is the memo type. Utilizing the 2-byte memo type, client applications can differentiate payment requests from payment intents.

Reviewer 2: ```suggestion
This is exactly identical to the aforementioned 0x0203 Destination With Payment Request ID Memo. The only difference is the memo type. Utilizing the 2-byte memo type, client applications can differentiate payment requests from payment intents.
```

# Drawbacks
[drawbacks]: #drawbacks

There is a limited number of memo types (65536). It likely won't be feasible to add more in the future. This proposal will utilize three more memo types that we won't be able to use for something else in the future.

> Additionally, adding three new memo types adds complexity to recoverable transaction history. Many client applications may not decide to utilize payment requests or intents but may wish to implement basic recoverable transaction history. Fortunately, this proposal cements the idea that the sender and destination memos are extended by the payment request ID and payment intent ID variations. It will take additional dev work, but MobileCoin clients and SDKs will be able to use this hierarchy to allow handling of the new memo types the same as the old ones. End-user facing applications may strictly write code to handle the basic sender and destination memo types. This code will always work regardless of any new features that they do not want being added in the future.

Reviewer 2: Nit: Capitalize RTH
```suggestion
Additionally, adding three new memo types adds complexity to Recoverable Transaction History. Many client applications may not decide to utilize payment requests or intents but may wish to implement basic recoverable transaction history. Fortunately, this proposal builds on the idea that the sender and destination memos are extended by the payment request ID and payment intent ID variations. It will take additional dev work, but MobileCoin clients and SDKs will be able to use this hierarchy to allow handling of the new memo types the same as the old ones. End-user facing applications may strictly write code to handle the basic sender and destination memo types. This code will always work regardless of any new features that they do not want being added in the future.
```

# Rationale and alternatives
[rationale-and-alternatives]: #rationale-and-alternatives

## Rationale

> The original intent of recoverable transaction history was that any two parties could recover the history of all transactions between each other. This has not been accomplished with the current implementation of payment request IDs. Since there is the ability for the recipient of a transaction to determine which payment request it relates to (if applicable), the sender of that transaction should also be able to do so. As of right now there is no way for the sender to know whether or not it relates to a payment request at all.

Reviewer 2: ```suggestion
The original intent of Recoverable Transaction History was that any two parties could recover the history of all transactions between each other. This has not been accomplished with the current implementation of payment request IDs. Since there is the ability for the recipient of a transaction to determine which payment request it relates to (when applicable), the sender of that transaction should also be able to do so. As of this writing there is no way for the sender to recover whether the payment relates to a request at all.
```

> Since recoverable transaction history supports payments requests, it follows that it should also support payment intents. This has many use-cases such as being able to send to a user that has not created or shared their public address with the prospective sender. This allows applications more options for things such as inviting new users or inintiating a new transaction history between (theretofore unintroduced) existing users.

Reviewer 2: ```suggestion
Since RTH supports payments requests, it follows that it should also support payment intents. This has many use-cases such as being able to invite a user that has not created or shared their public address with the prospective sender. This allows applications more options for things such as inviting new users or initiating a new transaction history between (theretofore unintroduced) existing users.
```
